applications/open-node-framework.md

```diff
@@ -0,0 +1,196 @@
+# W3F Open Grant Proposal
+
+* **Project Name:** Open Node Framework
+* **Team Name:** Phala Network
+* **Payment Address:** DAI at 0x50DC97D0345d61B4D096e15313d50b6506972e5F
+
+## Project Overview :page_facing_up: 
+
+### Overview
+
+**Open Node Framework** is an end-to-end DevOps (SRE) solution to deploy a Substrate node network for staking and other purposes. It's designed to address high availability, scalability, and flexibility, using the modern technology stack.
+
+### Project Details
+
+While developing a Substrate based blockchain, we found it challenging to operate and maintain different types of nodes. Validators (or collators in the future) require high availability and key security. Bootnodes and RPC nodes require high performance and flexible scalability. So Open Node Framework has the following design purpose:
+
+- High Availability: Add redundancy to nodes (validators & collators) for higher security and robustness
+- Flexibility: Integrate with any Substrate based blockchain easily
+- Multipurpose: Run validators, collators, full nodes, bootstrap nodes, or RPC nodes
+
+There are a few existing projects that can partially meet the above requirements. [Polkadot Secure Validator](https://github.com/w3f/polkadot-secure-validator) implements a validator setup for Polkadot and Kusama, but it doesn't come with a monitoring system. [Polkadot Deployer](https://github.com/w3f/polkadot-deployer) supports Docker-based network deployment and monitoring but is designed for bootstraping nodes. Both of them require some additional work to integrate with other Substrate blockchains. [Gantree](https://github.com/gantree-io/) is a W3F Grant funded Substrate DevOps framework supporting customized Substrate node. However it lacks the support of node HA, and like the other two projects, doesn't support cross-datacenter deployment, and is not built on modern Kubernetes stack.
+
+Open Node Framework wants to feature:
+
+- Multi IT infrastructure
+    - Native Kubernetes architecture
+    - Cross-datacenter / cloud providers deployment with Google Anthos
+    - Support multiple cloud providers with Terraform
+    - Scale up to a large node network easily
+- High availability
+    - Validator redundancy
```

Comment:
We considered the chance of double-signing in the design. In short, only one validator is supposed to be running at any time. We can prepare some archive nodes as the backup, and can convert a backup node to a validator quickly. In case of a validator corruption, a new validator can be spawn up with the session key injected, and recover functional.

There's the risk of network partition, where the controller may wrongly believe some validators are not accessible and provision the session key to multiple running validators causing equivocation.

In the long term, this should be handled gracefully with remote keystores (being implemented in Substrate side, but not functional now). In this case, the external keystore can be implemented with redundancy, but also controlled by a distributed lock infra to avoid double signing.

However in the short term, a walk-around to the network partition problem could be to send some messages from the node side to the blockchain (e.g. system.remark) to indicate the health condition, when the node suspect the network is partitioned. So the other side will stop provisioning session keys to avoid equivocation.

Reply:
> We can prepare some archive nodes as the backup, and can convert a backup node to a validator quickly. In case of a validator corruption, a new validator can be spawn up with the session key injected, and recover functional.

Archive nodes are not strictly needed. You can used pruned nodes, which usually have db sizes under 10gb, and can get provisioned from scratch in <5 minutes if need be. 

I'm less of a fan of injecting keys, as you need to _really_ make sure the first instance is down and won't come back up again. Right now 99% of slashing has happened because of session keys being duplicated. 

I'm not sure that some blockchain solution (ie remarks) works well for determining liveness or uptime- instead some type of service discovery might be a better approach.

And then perhaps have the keystore in a mountable volume only attached to one instance at a time.

Or just not backing up keys, and assuming 2 sessions of downtime are the next best thing.

This is all until remote signing is more feasible that is. In the long term there would be remote keystores with double signing protection.

Reply:
https://github.com/protofire/polkadot-failover-mechanism

This for example was another type of approach to this that a previous grantee made. Might be useful to dig into this as well.

Reply:
Looks a better solution than posting to the blockchain. It utilizes 2/3 voting and halts when the threshold cannot meet, thus can resist 1/3 failure. It's a clever and clear solution.